### ISSUE TITLE
Selectors issue in fluent-react

### ISSUE THREAD

**Issue by Alserda** [projectfluent/fluent.js#323]
Hi,

The syntax as documented in https://projectfluent.org/fluent/guide/selectors.html doesn't seem to work properly with the React bindings.

I'm using an approach identical to the `redux-async` example.  

**.ftl:**
```
unread-emails =
    You have { $emails_count ->
        [0] no unread emails
        [one] one unread email
       *[other] { $emails_count } unread emails
    }.
```

**.tsx:**
```
<Localized id='unread-emails' $emails_count={3}>
  <p>{'Amount of emails: {$emails_count}'}</p>
</Localized>
```

**Output:**
```
<p>Amount of emails: {$emails_count}</p>
```

I would also expect the fallback to make interpolate the `$emails_count` variable.

If I use the syntax `unread-emails = { $emails_count } unread emails` inside my .ftl file, it works fine. Same goes for the functions as used in the _hello-world_ example. 

My environment:
react - 16.6.3
react - 16.6.3
fluent - 0.9.1
fluent-langneg - 0.1.0
fluent-react - 0.8.2
typescript - 3.2.1

**Comment by stasm** [comment]
Thank you for filing this issue. Your FTL snippet looks fine; I'm surprised to learn that it doesn't work. I'll try to investigate it this week.

> I would also expect the fallback to make interpolate the $emails_count variable.

I've heard this request a few times (the last time was #280), and I understand that this would be helpful. My biggest fear right now is that adding this feature would result in hiding most of localization bugs from the developer. I think perhaps implementing this feature as only available in production would be one way to mitigate my concern. Another solution would involve adding good support for pseudolocales to `fluent-react`, which would make localizability bugs easy to discover during testing. (I'd like to do it anyways.) I'll look into both of these approaches. I'm also open to other ideas!

**Comment by stasm** [comment]
I tried adding the `unread-emails` message to the `hello-world` example. I also made sure to install the exact same versions of the dependencies, except for TypeScript, which examples don't use.

<details>
<summary>(Click to see the diff)</summary>

```diff
diff --git a/fluent-react/examples/hello-world/package.json b/fluent-react/examples/hello-world/package.json
index d17b7bba..0e03da7a 100644
--- a/fluent-react/examples/hello-world/package.json
+++ b/fluent-react/examples/hello-world/package.json
@@ -2,21 +2,21 @@
   "name": "fluent-react-example-hello-world",
   "version": "0.1.0",
   "private": true,
   "scripts": {
     "start": "parcel serve public/index.html",
     "build": "parcel build public/index.html"
   },
   "dependencies": {
-    "fluent": "^0.8.0",
-    "fluent-langneg": "^0.1.0",
-    "fluent-react": "file:../../",
-    "react": "^16.3.2",
-    "react-dom": "^16.3.3"
+    "fluent": "0.9.1",
+    "fluent-langneg": "0.1.0",
+    "fluent-react": "0.8.2",
+    "react": "16.6.3",
+    "react-dom": "16.6.3"
   },
   "devDependencies": {
     "parcel-bundler": "^1.9.7"
   },
   "engines": {
     "node": ">=8.9.0",
     "browsers": "Firefox >= 57"
   }
diff --git a/fluent-react/examples/hello-world/src/App.js b/fluent-react/examples/hello-world/src/App.js
index 6632300c..4a3a43d3 100644
--- a/fluent-react/examples/hello-world/src/App.js
+++ b/fluent-react/examples/hello-world/src/App.js
@@ -7,11 +7,14 @@ export default function App() {
       <Localized id="title">
         <h1>Hello, world!</h1>
       </Localized>
       <Localized id='today-is' $date={new Date()}>
         <p>
           {'Today is { DATETIME($date, month: "long", day: "numeric") }.'}
         </p>
       </Localized>
+      <Localized id='unread-emails' $emails_count={1}>
+        <p>{'Amount of emails: {$emails_count}'}</p>
+      </Localized>
     </div>
   );
 }
diff --git a/fluent-react/examples/hello-world/src/l10n.js b/fluent-react/examples/hello-world/src/l10n.js
index 1393707d..5e2dc07f 100644
--- a/fluent-react/examples/hello-world/src/l10n.js
+++ b/fluent-react/examples/hello-world/src/l10n.js
@@ -4,16 +4,22 @@ import { negotiateLanguages } from 'fluent-langneg/compat';
 const MESSAGES_ALL = {
   'pl': `
 title = Witaj świecie!
 today-is = Dziś jest { DATETIME($date, month: "long", day: "numeric") }.
   `,
   'en-US': `
 title = Hello, world!
 today-is = Today is { DATETIME($date, month: "long", day: "numeric") }.
+unread-emails =
+    You have { $emails_count ->
+        [0] no unread emails
+        [one] one unread email
+       *[other] { $emails_count } unread emails
+    }.
   `,
 };
 
 export function* generateBundles(userLocales) {
   // Choose locales that are best for the user.
   const currentLocales = negotiateLanguages(
     userLocales,
     ['en-US', 'pl'],
```

</details>


When I launched the example, everything worked as expected. Which suggests to me that the reason why you're not seeing the expected behavior is something else than the use of the select expression. I haven't yet had a chance to use `fluent-react` with TypeScript myself, but I've helped set it up for the [Common Voice](https://github.com/mozilla/voice-web) website and it worked fine. I'd be surprised if TypeScript was causing the problem here. I suspect it's something else.

To take a step back: for some reason the `<Localized id='unread-emails'>` component is not able to find the translated message. This usually happens when (a) the message is missing from all locales, or when (b) it's written in a way that makes the Fluent parser not see it. I think we can rule out (b) because it's clear that this is proper Fluent; just to make sure, I tested it locally and also in the [Fluent Playground](http://projectfluent.org/play/). So I think we can narrow it down to the the `<LocalizationProvider>` not finding that message at all.

I know that you mentioned that other messages worked fine:

> If I use the syntax unread-emails = { $emails_count } unread emails inside my .ftl file, it works fine. Same goes for the functions as used in the hello-world example.

Would you mind checking if this is the case again? Do you use more than one .ftl file? Can I see the code which creates the `bundles` iterable for the `<LocalizationProvider>`? Perhaps we will be able to find clues there. Thanks!

**Comment by Alserda** [comment]
Thank you for your extensive research on trying to find the issue. The changes that you've made inside of the example seem identical to my project. I'll try and re-produce this inside of a project that I can publicly share, and update my findings here. Stay tuned.

**Comment by Alserda** [comment]
Hi, I created a public repository with my source code. What I basically did was copy the source code from my private project. Both are identical, as are the packages.

My findings are that it works flawlessly after I copied it, so that didn't make sense to me. I then tried the same translation (`unread-emails`) in my private repositories, and this worked without any issues either.

I used my Macbook for this proof of concept, the issue I encountered was on a Windows machine. So what I'm going to do now is check if my proof of concept works properly on that machine. If so - I have no idea what caused the issue. 

If you're interested in my implementation of your package, here it is: https://github.com/Alserda/fluent-selectors-323. :)

**Comment by Alserda** [comment]
I discovered some new things, the environment that I posted along with the issue (apparently) weren't the versions that I was actually using. I was using:
```
"fluent": "0.8.1",
"fluent-langneg": "0.1.0",
"fluent-react": "0.8.1",
"react": "16.4.1",
"react-dom": "16.4.1"
```
I thought I posted this issue after updating to the environment that the initially provided one. Typing `npm update` is hard 😢. But the reason that I updated my packages were related to this bug.

What I did discover with all of this, is that the issue that caused me to write this issue has something to do with the combination of `fluent 0.8.1` and a Windows environment. The issue can be reproduced with my public repository on a windows machine.

Updating the `fluent` library to (at least) 0.9.1 solves my Windows-issue. Version 0.10.0 works good too, so this issue seems to have been resolved in the latest versions 😃

**Comment by zbraniecki** [comment]
thank you for reporting it! Closing :)

**Comment by stasm** [comment]
Thanks for digging deeper into this, @Alserda. I'm happy to hear that updating to a newer version of `fluent` fixed the issue.

I still wanted to get to the bottom of this, in case there was an OS-specific bug hiding somewhere in the code. I cloned your repo (thanks for preparing it!) in `cmd.exe`, ran `npm install` and `npm start`. I still wasn't able to reproduce the bug, however.

What I did discover, however, is a different bug, possibly related. `fluent-react` [0.8.2](https://github.com/projectfluent/fluent.js/blob/master/fluent-react/CHANGELOG.md#fluent-react-082-october-25-2018) added a small feature to allow using strings as children of `<Localized>`. In 0.8.1 and before, there always needed to be an element between `<Localized>` and the text content.

I noticed that your code depends on `fluent-react` ^0.8.1, which makes `npm install` install 0.8.3, so the bug is not visible. When I manually force-installed 0.8.1, I could only see a blank page. Curiously, the error console was empty. If you want to continue to use 0.8.1, I'd recommend wrapping `{children}` in a `<span>` here:

https://github.com/Alserda/fluent-selectors-323/blob/a84f6234c84f98663a552d309a52876040604438/src/components/Translated.tsx#L27

Or you could pin your dependency to 0.8.3. Looking back, I'm sorry that this feature shipped as a point update, rather than as a 0.9.0, which would have been more appropriate.

**Comment by Alserda** [comment]
I guess it's a good learning to actually pin the versions instead of allowing minor or patch versions to prevent confusion. 

I seem to have modified that `Translated.tsx` file and committed that by accident. In the actual private repository I wrapped `{children}` with a `<React.Fragment>` instead of a `<span>`, to add that feature to allow using strings in `0.8.1`. I think my version used `0.8.1` because I initially installed it with that version without ever updating the packages. 

Mystery solved 😄 - Thanks for maintaining this project. It's a real refreshment & fun to work with it.
